New operator that generates navmesh for selected objects

theanine3D opened the pull request:
This PR adds a new operator ("Generate Hubs Navigation Mesh") to the addon. The operator is accessible via the 3D viewport's "Object" menu while a mesh object is selected in object mode.

The operator is pretty straightforward - it makes a bmesh copy of the original object, checks the normals of all the faces, and then removes any that are too steep to walk on. Then it cleans up the result a bit (limited dissolve, triangulate, remove loose verts). The result is a lower-poly, more optimized mesh than the original model.

The operator is fast and creates an instant result to work with. In many cases the result needs only minimal editing afterward. The operator also adds the navmesh Hubs component to the newly generated object as well.

![navmesh-gen-1](https://user-images.githubusercontent.com/88953117/214867260-7b595461-8682-486c-b121-534c3e22775a.gif)

It currently has one limitation - it currently fails at generating _connected_ staircase slopes, because the operator removes any steep faces, including the 90 degree vertical geometry in stairs. It will correctly grab the top (flat) face of each stair, but it won't connect them, so the user will still need to edit the navmesh afterward in edge cases like that. See GIF below. However, it only takes a few seconds of manual editing to turn these unconnected faces into a usable navmesh slope.

![stair-example](https://user-images.githubusercontent.com/88953117/214868972-86bb2f4b-0741-4d2d-8428-bd7e2654437e.gif)

At the moment, the operator also adds a new material to the resulting navmesh object (if the material doesn't already exist), with a bright neon pink color, to make it easy to distinguish the navmesh in the 3D viewport from the original object. I added this functionality because otherwise it's too easy to accidentally select the original object when you're trying to edit your navmesh afterward. I tried to pick a color that is less likely to already be present in most scenes. Because navmeshes aren't actually rendered once exported to Hubs, the navmesh material is just there to make editing in Blender easier, with hopefully no impact to performance.

keianhzo commented on addons/io_hubs_addon/components/ui.py:
```diff
@@ -188,6 +188,7 @@ def object_menu_addition(self, context):
     layout.separator()
     op = layout.operator("wm.migrate_hubs_components")
+    op = layout.operator("object.generate_hubs_navmesh")
     op.is_registration = False
```
I believe this setting belongs to the previous operator.

keianhzo commented on addons/io_hubs_addon/components/operators.py:
```diff
@@ -433,6 +442,79 @@ def invoke(self, context, event):
+            return {'FINISHED'}
+
+        bm = bmesh.new()
+        bm.from_mesh(obj.data)
```
You probably need to apply transforms here so the actual geometry is transformed. Something like:
```python
        bpy.ops.object.transform_apply(location=False, rotation=True, scale=True)
```

keianhzo replied:
Maybe also applying operators could be a good thing to do as the user probably expects the nav-mesh generation to happen on the resulting mesh. nav-mesh generation for surfaces?

keianhzo commented on addons/io_hubs_addon/components/operators.py:
```diff
@@ -433,6 +442,79 @@ def invoke(self, context, event):
+    def execute(self, context):
+        import bmesh
+
+        obj = bpy.context.active_object
```
Would be great if we can generate a nav mesh from multiple selected objects instead of only the active one

keianhzo reviewed (changes requested):
This looks really good! Thanks for the contribution. I have left a couple of comments. We can keep the discussion open here to refine this.

theanine3D commented:
I've implemented all of the above requested changes/suggestions. However, the solidify / remesh combination has issues.

I added code for it but my tests with the Remesh modifier did not produce satisfactory results. So I left that part of the code commented. You can uncomment lines 493-495, 499-502, and 506-512, if you'd like to play around with it and see if maybe you can get a better result. My tests kept resulting in missing faces, large gaps in teh resulting model, large chunks of stairs still missing, etc.

All the other code changes work fine though. The updated code now supports generating 1 navmesh from multiple selected objects. The transforms are applied beforehand as well, and if any objects are multi-user objects, they're forced to become single-user before transforms are applied.

keianhzo commented:
Humm I didn't know that that addon already existed, I missed that in my search. but this works pretty good and it's not hard to set up. Might be good to give this a try to provide a fully functional solution.
